Added guide for using raw JSON REST requests with low-level client.

Djcarrillo6 opened the pull request:
### Description
Adds a new guide that demonstrates how to interact with the OpenSearch REST API using the OpenSearch .NET low-level client by executing raw JSON REST requests. It covers a number of operations such as setting up an OpenSearch instance, indexing document data, and executing CRUD operations (CREATE, READ, UPDATE and DELETE) on documents in the index.

### Issues Resolved
Resolves issue [403](https://github.com/opensearch-project/opensearch-net/issues/403)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented on guides/using-json-rest-requests.md:
```diff
@@ -0,0 +1,135 @@
+# Using Raw JSON REST Requests
```
Let's try to match the format in https://github.com/opensearch-project/opensearch-py/blob/main/guides/json.md?

Djcarrillo6 replied:
@dblock I don't think this repo has a sample directory like the Python client does. Should I create a sample dir in this .NET client repo & put it there? Or did you mean you want me to put the sample somewhere else as in another repo?

dblock commented on guides/using-json-rest-requests.md:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+private class CreatePitParams : RequestParameters<CreatePitParams>
```
This looks unused.

dblock commented on guides/using-json-rest-requests.md:
```diff
@@ -0,0 +1,135 @@
+
+var index = "movies";
+
+await client.IndexManyAsync(
```
For a pure JSON example you can't use `IndexManyAsync`.

dblock reviewed (commented):
Good start.

1. Remove any usage of higher level classes that aren't raw request/response.
2. Write a working sample and put it into `samples` so we can make sure this code works.

Djcarrillo6 commented:
> Good start.
> 
> 1. Remove any usage of higher level classes that aren't raw request/response.
> 2. Write a working sample and put it into `samples` so we can make sure this code works.

@dblock @Xtansia If I'm not mistake, this .NET client repo does not currently have a "samples" directory like the Python client does. Should I create that new dir here and place my working sample in that location? Or should I just attach the sample in the same "guides" directory where the MD files live?

dblock commented:
> @dblock @Xtansia If I'm not mistaken, this .NET client repo does not currently have a "samples" directory like the Python client does. Should I create that new directory here and place my working sample in that location? Or should I just attach the sample in the same "guides" directory where the MD files live?

Please create one!

Djcarrillo6 commented:
@dblock I added the new sample directory and added a working sample for my guide. I also change the name of guide file to match what the Python client [guide](https://github.com/opensearch-project/opensearch-py/blob/main/guides/json.md) used.

Djcarrillo6 commented:
Hey @Xtansia,  would you like me to add a link to this guide in the `USER_GUIDE.md` file? If so, which section to you think is best with respect to the guide's link placement.

Xtansia reviewed (changes requested):
@Djcarrillo6 A few points:
* Rather than using `new CancellationToken()` you should use `CancellationToken.None`
* The sample code should get added to a new project so that it actually gets compiled
* As far as linking the guide, I think we should add an `Advanced Features` section like some of the other clients have: https://github.com/opensearch-project/opensearch-py/blob/main/USER_GUIDE.md#advanced-features

Djcarrillo6 commented:
> @Djcarrillo6 A few points:
> 
> * Rather than using `new CancellationToken()` you should use `CancellationToken.None`
> * The sample code should get added to a new project so that it actually gets compiled
> * As far as linking the guide, I think we should add an `Advanced Features` section like some of the other clients have: https://github.com/opensearch-project/opensearch-py/blob/main/USER_GUIDE.md#advanced-features

@Xtansia what would adding the sample to a new project entail? Could I just add the sample file to the `src/OpenSearch.Net` directory?

> @Djcarrillo6 A few points:
> 
> * Rather than using `new CancellationToken()` you should use `CancellationToken.None`
> * The sample code should get added to a new project so that it actually gets compiled
> * As far as linking the guide, I think we should add an `Advanced Features` section like some of the other clients have: https://github.com/opensearch-project/opensearch-py/blob/main/USER_GUIDE.md#advanced-features

@Xtansia Can you detail how would I would correctly add this sample to a new project within the repo?

Xtansia commented:
@Djcarrillo6 You can create and add a new project with the `dotnet` CLI:
```bash
dotnet new console -o samples/Samples
dotnet sln add samples/Samples
```

Xtansia commented on USER_GUIDE.md:
```diff
@@ -303,3 +303,9 @@ var client = new OpenSearchLowLevelClient(config);
 Note the main difference here is that we are instantiating an `OpenSearchLowLevelClient` rather than `OpenSearchClient`, and `ConnectionConfiguration` instead of `ConnectionSettings`.
+
+
+## Advanced Features
+
+- [Making Raw JSON Requests](guides/json.md)
+  - [Sample Code](/samples/Samples/Program.cs)
```
You do not need the `[Sample Code]` line here. Additionally there should be a entry added to the table of contents at the top of this file for the `Advanced Features`

Xtansia commented on samples/Samples/Program.cs:
```diff
@@ -0,0 +1,49 @@
+﻿using OpenSearch.Client;
```
This needs the license header added:
```csharp
/* SPDX-License-Identifier: Apache-2.0
*
* The OpenSearch Contributors require contributions made to
* this file be licensed under the Apache-2.0 license or a
* compatible open source license.
*/
```

Xtansia commented on samples/Samples/Samples.csproj:
```diff
@@ -0,0 +1,14 @@
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
```
Add a `<IsPackable>False</IsPackable>` here to make sure it won't get accidentally packed for package publishing

Xtansia commented on samples/Samples/Samples.csproj:
```diff
@@ -0,0 +1,14 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="OpenSearch.Client" Version="1.*" />
```
This should be a project reference rather than a package reference, so that the sample code will be forced to reflect changes in the library.

Xtansia commented on guides/json.md:
```diff
@@ -0,0 +1,52 @@
+  - [DELETE](#delete)
+
+# Making Raw JSON REST Requests
+OpenSearch exposes a REST API that you can use to interact with OpenSearch. The OpenSearch .NET client provides a low-level API that allows you to send raw JSON requests to OpenSearch. This API is useful if you want to use a feature that is not yet supported by the OpenSearch .NET client, but it supported by the OpenSearch REST API.
```
You can link to the sample code from here, like the python guide does.

Additionally I think it'd make more sense to use the same example actions that the python guide does: https://github.com/opensearch-project/opensearch-py/blob/main/guides/json.md
https://github.com/opensearch-project/opensearch-py/blob/main/samples/json/hello.py

Xtansia commented on guides/json.md:
````diff
@@ -0,0 +1,69 @@
+```
+
+# PUT
+The following example creates a document via `PUT /movies/_doc/{id}` with a request body that specifies the document to create.
````
This and the rest of the descriptions are no-longer accurate

dblock commented on guides/json.md:
```diff
@@ -0,0 +1,69 @@
+            number_of_shards = 4
+        }
+    }
+};
```
Is there a JSON/string version of this? Not sure what's best for this example, maybe both?

dblock reviewed (commented):
Let's make a subfolder for the sample and call it "RawJsonSample"?

Djcarrillo6 commented:
@Xtansia I added the new folder structure & converted the sample to use raw JSON to match the guide implementation as @dblock requested.

Xtansia reviewed (changes requested):
If you're using a string body, you should use `PostData.String` rather than `PostData.Serializable`. The guide examples need to be updated to match the sample code. It'd be good to add a section to the guide informing users about the different types of `PostData` they can use with examples.

Djcarrillo6 commented:
> If you're using a string body, you should use `PostData.String` rather than `PostData.Serializable`. The guide examples need to be updated to match the sample code. It'd be good to add a section to the guide informing users about the different types of `PostData` they can use with examples.

@Xtansia I made the change to PostData.String and added a section to the guide informing users about the different `PostData` they can use with examples. Let me know if you'd like anything changed

dblock commented:
Looking at how this is done in the .NET client, would aliasing `LowLevel` to `.http` make sense to make the API more digestable? Similar to https://github.com/opensearch-project/opensearch-py/pull/544.
